core/src/main/scala/chisel3/properties/Property.scala

```diff
@@ -40,38 +67,72 @@ private[chisel3] trait PropertyType[T] {
+      def getPropertyType(value: Option[T]): fir.PropertyType = getType(value)
+      def convert(value:         T):         fir.Expression = getExpression(value)
+    }
+
   @implicitAmbiguous("unable to infer Property type. Please specify it explicitly in square brackets on the LHS.")
-  implicit val intPropertyTypeInstance = new PropertyType[Int] {
-    override def getPropertyType(_value: Option[Int]): ir.PropertyType = ir.IntegerPropertyType
-  }
+  implicit val intPropertyTypeInstance = makeSimple[Int](_ => fir.IntegerPropertyType, fir.IntegerPropertyLiteral(_))
 
   @implicitAmbiguous("unable to infer Property type. Please specify it explicitly in square brackets on the LHS.")
-  implicit val longPropertyTypeInstance = new PropertyType[Long] {
-    override def getPropertyType(_value: Option[Long]): ir.PropertyType = ir.IntegerPropertyType
-  }
+  implicit val longPropertyTypeInstance = makeSimple[Long](_ => fir.IntegerPropertyType, fir.IntegerPropertyLiteral(_))
 
   @implicitAmbiguous("unable to infer Property type. Please specify it explicitly in square brackets on the LHS.")
-  implicit val bigIntPropertyTypeInstance = new PropertyType[BigInt] {
-    override def getPropertyType(_value: Option[BigInt]): ir.PropertyType = ir.IntegerPropertyType
-  }
+  implicit val bigIntPropertyTypeInstance =
+    makeSimple[BigInt](_ => fir.IntegerPropertyType, fir.IntegerPropertyLiteral(_))
 
   @implicitAmbiguous("unable to infer Property type. Please specify it explicitly in square brackets on the LHS.")
-  implicit val classPropertyTypeInstance = new PropertyType[ClassType] {
-    override def getPropertyType(value: Option[ClassType]): ir.PropertyType = ir.ClassPropertyType(value.get.name)
-  }
+  implicit val classPropertyTypeInstance = makeSimple[ClassType](
+    value => fir.ClassPropertyType(value.get.name),
+    value => fir.StringPropertyLiteral(value.name)
+  )
 
-  implicit val stringPropertyTypeInstance = new PropertyType[String] {
-    override def getPropertyType(value: Option[String]): ir.PropertyType = ir.StringPropertyType
-  }
+  implicit val stringPropertyTypeInstance =
+    makeSimple[String](_ => fir.StringPropertyType, fir.StringPropertyLiteral(_))
+
+  implicit val boolPropertyTypeInstance =
+    makeSimple[Boolean](_ => fir.BooleanPropertyType, fir.BooleanPropertyLiteral(_))
 
-  implicit val boolPropertyTypeInstance = new PropertyType[Boolean] {
-    override def getPropertyType(value: Option[Boolean]): ir.PropertyType = ir.BooleanPropertyType
+  implicit def propertyTypeInstance[T](implicit pte: PropertyType[T]) = new PropertyType[Property[T]] {
+    type Type = T
+    override def getPropertyType(value: Option[Property[T]]): fir.PropertyType = pte.getPropertyType(None)
+    override def convert(value:         Underlying, ctx: ir.Component, info: SourceInfo): fir.Expression =
+      ir.Converter.convert(value, ctx, info)
+    type Underlying = ir.Arg
+    override def convertUnderlying(value: Property[T]) = value.ref
```

Comment:
I get why this is here, but I think we should think about if this is the only way to accomplish the goal of making `Property.apply` work for the likes of both `Property[Seq[BigInt]]` and `Seq[Property[BigInt]]`. It using overloading so bad since we only have the 2 collection types? I think we will just need 3 in that case.

What worries me about this is it makes it legal to write things like `Property[Seq[Property[Property[Property[BigInt]]]]]` and I'm not sure what that means.

Reply:
I don't think we should sandbag on this because overall this PR is a amazing and a huge step forward.
Please leave a comment explaining why this is here with maybe a TODO to see if there's either a way to remove the need for `PropertyType[Property[T]]` or at least disallow weird `Property[Property[T]]` or better yet disallow mutually recursive types like `Property[Seq[Property[Seq[Property[T]]]]]` (ideally disallow more than 1 `Property[_]` in the type if at all possible).

We can try to "fix" this later, let's definitely get this PR in.